libraries/stdlib/jvm/src/kotlin/text/StringNumberConversionsJVM.kt

```diff
@@ -239,35 +239,215 @@ public fun String.toBigDecimalOrNull(): java.math.BigDecimal? =
+
+        // Test for [xX] to see if we truly have a hex string
+        if (s[start].code or 0x20 == 'x'.code) {
+            start++
+
+            // Look for hex digits after the 0x prefix
+            var checkpoint = start
+            while (start <= end) {
+                val d = s[start]
+                if (d.isAsciiDigit() || d.isHexLetter()) {
+                    start++
+                } else {
+                    break
+                }
+            }
+            // Check if we found 0x*****, otherwise, the hex number might be of the
+            // form 0x.*******
+            hasIntegerPart = checkpoint != start
+
+            // A hex string must have an exponent, the string is invalid if we only found an
+            // integer part
+            if (start > end) return false
+
+            if (s[start] == '.') {
+                start++
+
+                // Look for hex digits for the fractional part
+                checkpoint = start
+                while (start <= end) {
+                    val d = s[start]
+                    if (d.isAsciiDigit() || d.isHexLetter()) {
+                        start++
+                    } else {
+                        break
+                    }
+                }
+
+                // Did we find a fractional part?
+                hasFractionalPart = checkpoint != start
+            }
+
+            // A string must have an integer part, or a fractional part, or both
+            if (!hasIntegerPart && !hasFractionalPart) return false
+
+            // A hex string must have an exponent, the string is invalid if we only found an
+            // integer and/or fractional part
+            if (start > end) return false
+
+            isHex = true
+        } else {
+            // Rewind the 0 we just parsed to make things easier below and try to parse a non-
+            // hexadecimal string representation of a float
+            start--
+        }
+    }
+
+    // Parse a non-hexadecimal representations
+    if (!isHex) {
+        // Look for digits before the decimal separator, if any
+        var checkpoint = start
+        while (start <= end && s[start].isAsciiDigit()) start++
+
+        // If there's no integer part, the float might be of the form .1234
+        hasIntegerPart = checkpoint != start
+
+        // A non-hexadecimal representation only needs an integer part, we can stop here
+        if (start > end) return hasIntegerPart
+
+        if (s[start] == '.') {
+            start++
+
+            // Look for the fractional part
+            checkpoint = start
+            while (start <= end && s[start].isAsciiDigit()) start++
+
+            // Did we find a fractional part?
+            hasFractionalPart = checkpoint != start
+        }
+
+        // A string must have an integer part, or a fractional part, or both
+        if (!hasIntegerPart && !hasFractionalPart) {
+            // Special case non-finite constants
+            val constant = when (end) {
```

Comment:
For the sake of readability, I suggest outlining special-case handling to a separate function as well.
`LengthOfXXX` could then be defined inside the function (They are not lengths, so it would be better to make their scope as narrow as possible).

> For the sake of readability

Don't get me wrong: the function is well written, there's just too much context to keep in mind while reading the code.